Add useq linter

dhaus67 opened the pull request:
Adding a new linter called [useq](https://github.com/dhaus67/useq).

The linter is checking formatting arguments and replacing manually crafted escapes (such as `fmt.Printf("Hello my name is \"%s\"", name)` with the `%q` formatting directive (resulting in `fmt.Printf("Hello my name is %q", name)`.


Some unrelated changes have been made to the `go.mod` since this seems to be the first module that introduces `go1.23`. If we want to make this change separately and keep `go 1.22.1` for the time being, I'm happy to downgrade the go version used in [the useq](https://github.com/dhaus67/useq) package for the time being.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/golangci/golangci-lint?pullRequest=5033) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/golangci/golangci-lint?pullRequest=5033) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/golangci/golangci-lint?pullRequest=5033) it.</sub>

ldez commented:
In order for a pull request adding a linter to be reviewed, the linter and the PR must follow some requirements.

- [ ] The CLA must be signed

### Pull Request Description

- [ ] It must have a link to the linter repository.
- [ ] It must provide a short description of the linter.

### Linter

- [ ] It must not be a duplicate of another linter or a rule of a linter (the team will help to verify that).
- [ ] It must have a valid license (AGPL is not allowed), and the file must contain the required information by the license, ex: author, year, etc.
- [ ] It must use Go version <= 1.22
- [ ] The linter repository must have a CI and tests.
- [ ] It must use [`go/analysis`](https://golangci-lint.run/contributing/new-linters/).
- [ ] It must have a valid tag, ex: `v1.0.0`, `v0.1.0`.
- [ ] It must not contain `init()`.
- [ ] It must not contain `panic()`.
- [ ] It must not contain `log.Fatal()`, `os.Exit()`, or similar.
- [ ] It must not modify the AST.
- [ ] It must not have false positives/negatives (the team will help to verify that).
- [ ] It must have tests inside golangci-lint.

### The Linter Tests Inside Golangci-lint

- [ ] They must have at least one std lib import.
- [ ] They must have integration tests without configuration (default).
- [ ] They must have integration tests with configuration (if the linter has a configuration).

### `.golangci.next.reference.yml`

- [ ] The file `.golangci.next.reference.yml` must be updated.
- [ ] The file `.golangci.reference.yml` must NOT be edited.
- [ ] The linter must be added to the lists of available linters (alphabetical case-insensitive order).
    - `enable` and `disable` options
- [ ] If the linter has a configuration, the exhaustive configuration of the linter must be added (alphabetical case-insensitive order)
    - The values must be different from the default ones.
    - The default values must be defined in a comment.
    - The option must have a short description.

### Others Requirements

- [ ] The files (tests and linter) inside golangci-lint must have the same name as the linter.
- [ ] The `.golangci.yml` of golangci-lint itself must not be edited and the linter must not be added to this file.
- [ ] The linters must be sorted in the alphabetical order (case-insensitive) in the `lintersdb/builder_linter.go` and `.golangci.next.reference.yml`.
- [ ] The load mode (`WithLoadMode(...)`):
    - if the linter uses `goanalysis.LoadModeSyntax` -> no `WithLoadForGoAnalysis()` in `lintersdb/builder_linter.go`
    - if the linter uses `goanalysis.LoadModeTypesInfo`, it requires `WithLoadForGoAnalysis()` in `lintersdb/builder_linter.go`
- [ ] The version in `WithSince(...)` must be the next minor version (`v1.X.0`) of golangci-lint.
- [ ] `WithURL()` must contain the URL of the repository.

### Recommendations

- [ ] The file `jsonschema/golangci.next.jsonschema.json` should be updated.
- [ ] The file `jsonschema/golangci.jsonschema.json` must NOT be edited.
- [ ] The linter repository should have a readme and linting.
- [ ] The linter should be published as a binary (useful to diagnose bug origins).
- [ ] The linter repository should have a `.gitignore` (IDE files, binaries, OS files, etc. should not be committed)
- [ ] A tag should never be recreated.
- [ ] use `main` as the default branch name.

---

The golangci-lint team will edit this comment to check the boxes before and during the review.

The code review will start after the completion of those checkboxes (except for the specific items that the team will help to verify).

The reviews should be addressed as commits (no squash).

If the author of the PR is a member of the golangci-lint team, he should not edit this message.

**This checklist does not imply that we will accept the linter.**

ldez commented:
Could remove all the things that are not related directly to the PR?
- documentation changes
- Go version change (We need to use go1.22)
- etc.

dhaus67 commented:
@ldez thanks for the swift review!

I've cleaned up the PR and downgraded the Go version.

For couple of the checkboxes:
- License is included in the linter repository [here](https://github.com/dhaus67/useq/blob/main/LICENSE), do you also need a license header for each file?
- Added the required doc in the `.golangci.next.reference.yml` let me know what you think.
- The `.gitignore` for the linter itself is [here](https://github.com/dhaus67/useq/blob/main/.gitignore).

Let me know if there's more things to address, thanks again for the quick review and clear checklist :)

ldez commented:
- For the license: https://github.com/dhaus67/useq/blob/fe8ed4a03c8f1b4e857abd8a98967e532434eef6/LICENSE#L178-L201
- I know where is the `.gitignore`. The binary, resulting in a single `go build ./cmd/useq` is not excluded.
- the content of `.golangci.next.reference.yml` still does not follow the style of the other linters.

Note: Please, no squash :wink:

ldez commented:
This rule already exists inside golangci-lint: [gocritic](https://golangci-lint.run/usage/linters/#gocritic) ([sprintfQuotedString](https://go-critic.com/overview#sprintfquotedstring))

We don't accept duplicate rule/linter, so we will decline your PR.

dhaus67 commented:
@ldez thanks for pointing out the duplicate, however this only covers `fmt.Sprintf` calls while this linter would cover essentially any printf like methods.
Should this rather be extended in gocritic then if we don't want to add `useq` in golangci-lint?

bombsimon commented:
> Should this rather be extended in gocritic then if we don't want to add `useq` in golangci-lint?

Yes, filing a PR to improve the feature in `gocritic` would be the best option here!
